File: THCS/2023_2024/ThanhHoa_2324/Cau4/CAU4.cpp
#include <bits/stdc++.h>
#define int long long
using namespace std;

signed main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    
    // Mở file để đọc dữ liệu
    ifstream infile("CAU4.INP");
    // Kiểm tra xem file có mở thành công hay không
    if (!infile.is_open()) {
        cerr << "Không thể mở file CAU4.INP" << endl;
        return 1;
    }

    int n;
    infile >> n; // Đọc số lượng cặp

    int ans = 0;
    map<int, int> mp;
    map<pair<int, int>, int> mp2;

    for (int i = 1; i <= n; i++) {
        int a, b;
        infile >> a >> b; // Đọc các cặp số a và b

        int gcd_value = __gcd(a, b);
        int ff = a / gcd_value;
        int ss = b / gcd_value;

        ans += mp[a] - mp2[make_pair(ff, ss)];
        mp[a]++;
        mp2[{ff, ss}]++;
    }

    // Đóng file input
    infile.close();

    // In kết quả ra file output
    ofstream outfile("CAU4.OUT");
    if (!outfile.is_open()) {
        cerr << "Không thể mở file CAU4.OUT" << endl;
        return 1;
    }

    outfile << ans; // Ghi kết quả vào file
    outfile.close();

    return 0;
}

File: THCS/2023_2024/ThanhHoa_2324/Cau4/CAU4.py
from math import gcd
from collections import defaultdict

def main():
    # Mở file để đọc dữ liệu
    try:
        with open("CAU4.INP", "r") as infile:
            n = int(infile.readline().strip())  # Đọc số lượng cặp
            
            ans = 0
            mp = defaultdict(int)  # Map để lưu số lượng các giá trị a
            mp2 = defaultdict(int) # Map để lưu số lượng cặp (ff, ss)

            for _ in range(n):
                a, b = map(int, infile.readline().strip().split())  # Đọc các cặp số a và b

                gcd_value = gcd(a, b)
                ff = a // gcd_value
                ss = b // gcd_value

                ans += mp[a] - mp2[(ff, ss)]
                mp[a] += 1
                mp2[(ff, ss)] += 1

    except FileNotFoundError:
        print("Không thể mở file CAU4.INP")
        return

    # In kết quả ra file output
    try:
        with open("CAU4.OUT", "w") as outfile:
            outfile.write(str(ans))  # Ghi kết quả vào file
    except:
        print("Không thể mở file CAU4.OUT")

if __name__ == "__main__":
    main()
